NoteEditor - CardTemplatePreviewer - ArrayIndexOutOfBoundsException

https://couchdb.ankidroid.org/acralyzer/_design/acralyzer/index.html#/report-details/a01688a2-216f-4aa0-9af2-4d395f61fdc6

2.12.0beta3


```
0	--------- beginning of main
1	07-06 17:02:45.243 I/AnkiDroid(24389): CardBrowser:: Completed doInBackgroundSearchCards Successfully
2	07-06 17:02:45.382 I/AnkiDroid(24389): AnkiActivity::onStop - NoteEditor
3	07-06 17:02:45.386 I/AnkiDroid(24389): AnkiActivity::onDestroy - NoteEditor
4	07-06 17:02:45.937 I/AnkiDroid(24389): AnkiActivity::onPause - CardBrowser
5	07-06 17:02:45.953 I/AnkiDroid(24389): AnkiActivity::onCreate - NoteEditor
6	07-06 17:02:45.990 I/AnkiDroid(24389): onCollectionLoaded() Edit note activity successfully started with card id 1485680989833
7	07-06 17:02:45.997 I/AnkiDroid(24389): AnkiActivity::onStart - NoteEditor
8	07-06 17:02:46.001 I/AnkiDroid(24389): AnkiActivity::onResume - NoteEditor
9	07-06 17:02:46.011 I/AnkiDroid(24389): Disabling Extended Text UI
10	07-06 17:02:46.011 I/AnkiDroid(24389): Disabling Extended Text UI
11	07-06 17:02:46.288 I/AnkiDroid(24389): AnkiActivity::onStop - CardBrowser
12	07-06 17:02:46.291 I/AnkiDroid(24389): flush - Saving information to DB...
13	07-06 17:02:47.621 I/AnkiDroid(24389): Disabling Extended Text UI
14	07-06 17:02:47.631 I/AnkiDroid(24389): Disabling Extended Text UI
15	07-06 17:02:49.472 I/AnkiDroid(24389): NoteEditor:: Preview button pressed
16	07-06 17:02:49.474 I/AnkiDroid(24389): Saving instance
17	07-06 17:02:49.498 I/AnkiDroid(24389): AnkiActivity::onPause - NoteEditor
18	07-06 17:02:49.534 I/AnkiDroid(24389): AnkiActivity::onCreate - CardTemplatePreviewer
```

```
java.lang.RuntimeException: Unable to start activity ComponentInfo{com.ichi2.anki/com.ichi2.anki.CardTemplatePreviewer}: java.lang.ArrayIndexOutOfBoundsException: length=2; index=2
at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3448)
at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3595)
at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:83)
at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2147)
at android.os.Handler.dispatchMessage(Handler.java:107)
at android.os.Looper.loop(Looper.java:237)
at android.app.ActivityThread.main(ActivityThread.java:7814)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1075)
Caused by: java.lang.ArrayIndexOutOfBoundsException: length=2; index=2
at com.ichi2.libanki.Note.setField(Note.java:1)
at com.ichi2.anki.CardTemplatePreviewer.onCollectionLoaded(CardTemplatePreviewer.java:11)
at com.ichi2.anki.AnkiActivity.startLoadingCollection(AnkiActivity.java:4)
at com.ichi2.anki.CardTemplatePreviewer.onCreate(CardTemplatePreviewer.java:19)
at android.app.Activity.performCreate(Activity.java:7955)
at android.app.Activity.performCreate(Activity.java:7944)
at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1307)
at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3423)
... 11 more
```

--- a/AnkiDroid/src/main/java/com/ichi2/anki/CardTemplatePreviewer.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/CardTemplatePreviewer.java
@@ -19,6 +19,7 @@ package com.ichi2.anki;
 import android.os.Bundle;
 import android.view.View;
 
+import com.ichi2.anim.ActivityTransitionAnimation;
 import com.ichi2.libanki.Card;
 import com.ichi2.libanki.Collection;
 import com.ichi2.libanki.Models;
@@ -66,6 +67,7 @@ public class CardTemplatePreviewer extends AbstractFlashcardViewer {
                 mEditedModel = TemporaryModel.getTempModel(mEditedModelFileName);
             } catch (IOException e) {
                 Timber.w(e, "Unable to load temp model from file %s", mEditedModelFileName);
+                closeCardTemplatePreviewer();
             }
         }
 
@@ -74,7 +76,7 @@ public class CardTemplatePreviewer extends AbstractFlashcardViewer {
             mCurrentCard = getDummyCard(mEditedModel, mOrdinal);
             if (mCurrentCard == null) {
                 UIUtils.showSimpleSnackbar(this, R.string.invalid_template, false);
-                finishWithoutAnimation();
+                closeCardTemplatePreviewer();
             }
         }
 
@@ -89,8 +91,30 @@ public class CardTemplatePreviewer extends AbstractFlashcardViewer {
         super.onResume();
         if (mCurrentCard == null || mOrdinal < 0) {
             Timber.e("CardTemplatePreviewer started with empty card list or invalid index");
-            finishWithoutAnimation();
+            closeCardTemplatePreviewer();
         }
+    }
+
+
+    private void closeCardTemplatePreviewer() {
+        Timber.d("CardTemplatePreviewer:: closeCardTemplatePreviewer()");
+        setResult(RESULT_OK);
+        TemporaryModel.clearTempModelFiles();
+        finishWithAnimation(ActivityTransitionAnimation.RIGHT);
+    }
+
+
+    @Override
+    public void onBackPressed() {
+        Timber.i("CardTemplatePreviewer:: onBackPressed()");
+        closeCardTemplatePreviewer();
+    }
+
+
+    @Override
+    protected void onNavigationPressed() {
+        Timber.i("CardTemplatePreviewer:: Navigation button pressed");
+        closeCardTemplatePreviewer();
     }
 
 
@@ -136,6 +160,11 @@ public class CardTemplatePreviewer extends AbstractFlashcardViewer {
     @Override
     protected void onCollectionLoaded(Collection col) {
         super.onCollectionLoaded(col);
+        if ((mCurrentCard == null) && (mCardList == null)) {
+            Timber.d("onCollectionLoaded - incorrect state to load, closing");
+            closeCardTemplatePreviewer();
+            return;
+        }
         if (mCurrentCard == null) {
             mCurrentCard = new PreviewerCard(col, mCardList[mOrdinal]);
         }
@@ -150,7 +179,11 @@ public class CardTemplatePreviewer extends AbstractFlashcardViewer {
             Bundle noteFields = mNoteEditorBundle.getBundle("editFields");
             if (noteFields != null) {
                 for (String fieldOrd : noteFields.keySet()) {
-                    currentNote.setField(Integer.parseInt(fieldOrd), noteFields.getString(fieldOrd));
+                    // In case the fields on the card are out of sync with the bundle
+                    int fieldOrdInt = Integer.parseInt(fieldOrd);
+                    if (currentNote.getFields().length < fieldOrdInt) {
+                        currentNote.setField(fieldOrdInt, noteFields.getString(fieldOrd));
+                    }
                 }
             }
         }
@@ -171,7 +204,7 @@ public class CardTemplatePreviewer extends AbstractFlashcardViewer {
         }
         Note n = getCol().newNote(model);
         ArrayList<String> fieldNames = Models.fieldNames(model);
-        for (int i = 0; i < fieldNames.size(); i++) {
+        for (int i = 0; i < fieldNames.size() && i < n.getFields().length; i++) {
             n.setField(i, fieldNames.get(i));
         }
         try {

--- a/AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.java
@@ -154,6 +154,7 @@ public class NoteEditor extends AnkiActivity {
     public static final int REQUEST_ADD = 0;
     public static final int REQUEST_MULTIMEDIA_EDIT = 2;
     public static final int REQUEST_TEMPLATE_EDIT = 3;
+    public static final int REQUEST_PREVIEW = 4;
 
     private boolean mChanged = false;
     private boolean mTagsEdited = false;
@@ -936,7 +937,7 @@ public class NoteEditor extends AnkiActivity {
                 Bundle noteEditorBundle = new Bundle();
                 onSaveInstanceState(noteEditorBundle);
                 previewer.putExtra("noteEditorBundle", noteEditorBundle);
-                startActivityWithoutAnimation(previewer);
+                startActivityForResultWithoutAnimation(previewer, REQUEST_PREVIEW);
                 return true;
 
             case R.id.action_save:
@@ -1065,6 +1066,8 @@ public class NoteEditor extends AnkiActivity {
         } else {
             setResult(result);
         }
+        // ensure there are no orphans from possible edit previews
+        TemporaryModel.clearTempModelFiles();
         if (mCaller == CALLER_CARDEDITOR_INTENT_ADD) {
             finishWithAnimation(ActivityTransitionAnimation.NONE);
         } else {

